test.py

```diff
@@ -1,131 +1,175 @@
+    # and its next_fast_len is the same as the len(T).
+    # To this end, we choose 3 and 5
+
+    n_T = (3**2) * (5**3)
+    shape = next_fast_len(n_T)
+    if shape != n_T:
+        msg = (
+            "In this test, n_T value should be set to an odd value "
+            + "such that its next_fast_len must be the same"
+        )
+        raise ValueError(msg)
+
+    modules = utils.import_sdp_mods()
+    for mod in modules:
+        for n_Q in range(2, n_T + 1):
+            try:
+                Q = np.random.rand(n_Q)
+                T = np.random.rand(n_T)
+
+                ref = naive_sliding_dot_product(Q, T)
+                comp = mod.sliding_dot_product(Q, T)
+                npt.assert_allclose(comp, ref)
+
+            except Exception as e:
+                msg = f"Error in {mod.__name__}, with n_Q={n_Q} and n_T={n_T}"
+                print(msg)
+                raise e
+
+    return
+
+
+def test_sdp_case3():
+    # This tests cases where the length of `T` is even
+    # and its next_fast_len is not the same as
+    # the len(T) but longer. To this end, we choose
+    # factors 2, 7, 11, 13.
+
+    n_T = 2 * 7 * 11 * 13
+    shape = next_fast_len(n_T)
+    if shape <= n_T:
+        msg = (
+            "In this test, n_T value should be set to an even value "
+            + "such that its next_fast_len must be larger"
+        )
+        raise ValueError(msg)
+
+    modules = utils.import_sdp_mods()
+    for mod in modules:
+        for n_Q in range(2, n_T + 1):
+            try:
+                Q = np.random.rand(n_Q)
+                T = np.random.rand(n_T)
+
+                ref = naive_sliding_dot_product(Q, T)
+                comp = mod.sliding_dot_product(Q, T)
+                npt.assert_allclose(comp, ref)
+
+            except Exception as e:
+                msg = f"Error in {mod.__name__}, with n_Q={n_Q} and n_T={n_T}"
+                print(msg)
+                raise e
+
+    return
+
+
+def test_sdp_case4():
+    # This tests cases where the length of `T` is odd
+    # and its next_fast_len is not the same as
+    # the len(T) but longer. To this end, we choose
+    # factors 7, 11, 13.
+
+    n_T = 7 * 11 * 13
+    shape = next_fast_len(n_T)
+    if shape <= n_T:
+        msg = (
+            "In this test, n_T value should be set to an odd value "
+            + "such that its next_fast_len must be larger"
+        )
+        raise ValueError(msg)
+
+    modules = utils.import_sdp_mods()
+    for mod in modules:
+        for n_Q in range(2, n_T + 1):
+            try:
+                Q = np.random.rand(n_Q)
+                T = np.random.rand(n_T)
+
+                ref = naive_sliding_dot_product(Q, T)
+                comp = mod.sliding_dot_product(Q, T)
+                npt.assert_allclose(comp, ref)
+
+            except Exception as e:
+                msg = f"Error in {mod.__name__}, with n_Q={n_Q} and n_T={n_T}"
+                print(msg)
+                raise e
+
+    return
```

Comment:
I did not add any test for `oaconvolve` because of two reasons:

(1) Having large `len(T)` alone does not guarantee that the overlap-add part of `oaconvolve` is going to be triggered as that depends on both `len(T)` and `len(Q)`. 

(2) I am now testing for a wide range of len(Q), from 2 to len(T). So, I think overlap-add part is going to be triggered at some point.

Reply:
The [module sdp/scipy_oaconvolve_sdp.py](https://github.com/NimaSarajpoor/sliding_dot_product/blob/84837766959e989bd0589fc221be858d37b58153/sdp/scipy_oaconvolve_sdp.py#L9-L10) uses the oaconvolve function from scipy and I do not think we can "easily" figure out whether it, internally, fallbacks to fftconvolve or not unless we go through that source code and check the fallback function  Still, your concern should be addressed:

> Well, you can be 100% sure of this if you check test coverage...

 Because, if we later add something like PR #17, then it is important to know if all of its parts are tested.

Reply:
My mistake was that I saw the current modules... and said: "okay...no coverage is needed". That was a very bad mindset. I should have thought about keeping ourselves safe for the long term when other new modules are added.

Reply:
It's okay. You're not wrong. You made a rational decision at the time. Now, the circumstances are changing and so we should update our perspective accordingly. :)